rest/oidc_api_test.go

```diff
@@ -130,10 +205,162 @@ func TestGetOIDCCallbackURL(t *testing.T) {
+		require.NoError(t, err, "Error sending request")
+		require.Equal(t, http.StatusOK, response.StatusCode)
+
+		// Validate received token response
+		var receivedToken OIDCTokenResponse
+		require.NoError(t, err, json.NewDecoder(response.Body).Decode(&receivedToken))
+		assert.NotEmpty(t, receivedToken.SessionID, "session_id doesn't exist")
+		assert.Equal(t, wantUsername, receivedToken.Username, "name mismatch")
+		assert.Equal(t, wantTokenResponse.IDToken, receivedToken.IDToken, "id_token mismatch")
+		assert.Equal(t, wantTokenResponse.RefreshToken, receivedToken.RefreshToken, "refresh_token mismatch")
+		assert.Equal(t, wantTokenResponse.AccessToken, receivedToken.AccessToken, "access_token mismatch")
+		assert.Equal(t, wantTokenResponse.TokenType, receivedToken.TokenType, "token_type mismatch")
+		assert.True(t, wantTokenResponse.Expires >= receivedToken.Expires, "expires_in mismatch")
+
+		// Query db endpoint with bearer token
+		var responseBody db.Body
+		dbEndpoint := mockSyncGatewayURL + "/" + restTester.DatabaseConfig.Name
+		request, err = http.NewRequest(http.MethodGet, dbEndpoint, nil)
+		request.Header.Add("Authorization", receivedToken.IDToken)
+		response, err = http.DefaultClient.Do(request)
+		require.NoError(t, err, "Error sending request with bearer token")
+		require.Equal(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, err, json.NewDecoder(response.Body).Decode(&responseBody))
+		assert.Equal(t, restTester.DatabaseConfig.Name, responseBody["db_name"])
+	})
+
+	t.Run("check whether state is maintained when callback state is enabled", func(t *testing.T) {
+		providerGoogle := auth.OIDCProvider{
+			Name: providerNameGoogle, Issuer: issuerGoogle, ClientID: clientIDGoogle, IncludeAccessToken: true,
+			UserPrefix: providerNameGoogle, ValidationKey: &validationKeyGoogle, Register: true,
+			DiscoveryURI: issuerGoogle + wellKnownPath, DisableCallbackState: base.BoolPtr(false),
+		}
+		providers := auth.OIDCProviderMap{providerGoogle.Name: &providerGoogle}
+		options := auth.OIDCOptions{Providers: providers, DefaultProvider: &providerGoogle.Name}
+		restTesterConfig := RestTesterConfig{DatabaseConfig: &DbConfig{OIDCConfig: &options}}
+		restTester := NewRestTester(t, &restTesterConfig)
+		defer restTester.Close()
+		fakeSyncGateway := httptest.NewServer(restTester.TestPublicHandler())
+		defer fakeSyncGateway.Close()
+		mockSyncGatewayURL = fakeSyncGateway.URL
+
+		// Initiate OpenID Connect Authorization Code flow.
+		requestURL := fmt.Sprintf("%s/db/_oidc?provider=google&offline=true", mockSyncGatewayURL)
+		request, err := http.NewRequest(http.MethodGet, requestURL, nil)
+		require.NoError(t, err, "Error creating new request")
+		jar, err := cookiejar.New(nil)
+		require.NoError(t, err, "Error creating new cookie jar")
+		client := &http.Client{Jar: jar}
+		response, err := client.Do(request)
+		require.NoError(t, err, "Error sending request")
+		require.Equal(t, http.StatusOK, response.StatusCode)
```

Comment:
Would be valuable to add a test which invalidates or changes the state cookie, to test the error path.

Reply:
Added unit test to check the invalidate state scenario